core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -109,130 +109,175 @@ final class FiberRefs private (
-                childDepth: Int
-              ): Any =
-                (parentStack, childStack) match {
-                  case (
-                        StackEntry(parentFiberId, parentValue, parentVersion) :: parentAncestors,
-                        StackEntry(childFiberId, childValue, childVersion) :: childAncestors
-                      ) =>
-                    if (parentFiberId == childFiberId)
-                      if (childVersion > parentVersion) parentValue else childValue
-                    else if (childDepth > parentDepth)
-                      findAncestor(parentStack, parentDepth, childAncestors, childDepth - 1)
-                    else if (childDepth < parentDepth)
-                      findAncestor(parentAncestors, parentDepth - 1, childStack, childDepth)
-                    else
-                      findAncestor(parentAncestors, parentDepth - 1, childAncestors, childDepth - 1)
-                  case _ =>
-                    (ref.initial)
-                }
-
-              val ancestor = findAncestor(parentStack, parentDepth, childStack, childDepth)
-
-              val patch = ref.diff(ancestor, childValue)
+                findAncestor(initialFiberValue)(parentAncestors, parentDepth - 1, childAncestors, childDepth - 1)
+          }
+      }
 
-              val oldValue = parentStack.head.value
-              val newValue = ref.join(oldValue, ref.patch(patch)(oldValue))
+    val parentFiberRefs = self.fiberRefLocals
+    val childFiberRefs  = that.fiberRefLocals
 
-              if (oldValue == newValue) parentFiberRefs
-              else {
-                val newEntry = parentStack match {
-                  case StackEntry(parentFiberId, _, parentVersion) :: tail =>
-                    if (parentFiberId == fiberId)
-                      Value(::(StackEntry(parentFiberId, newValue, parentVersion + 1), tail), parentDepth)
-                    else
-                      Value(::(StackEntry(fiberId, newValue, 0), parentStack), parentDepth + 1)
-                }
-                parentFiberRefs.updated(ref, newEntry)
-              }
+    val newFiberRefLocals = childFiberRefs.foldLeft(parentFiberRefs) { case (parentFiberRefs, (ref, childStack)) =>
+      if (childStack.fiberId == fiberId) parentFiberRefs
+      else {
+        val childFiberRef          = ref.asInstanceOf[FiberRef[Any]]
+        val childFiberInitialValue = childFiberRef.initial
+        val childStackCurrentValue = childStack.value
+
+        parentFiberRefs.getOrElse(childFiberRef, null) match {
+          case null =>
+            val emptyChildStack = childStackCurrentValue == childFiberInitialValue
+            if (emptyChildStack) parentFiberRefs
+            else
+              parentFiberRefs.updated(
+                childFiberRef,
+                FiberRefStack.init(fiberId, childFiberRef.join(childFiberInitialValue, childStackCurrentValue))
+              )
+          case parentStack =>
```

Comment:
In order to make the code compile after removing wildcards from the `FiberRefStack` methods / arguments

```suggestion
          case parentStack: FiberRefStack[Any] =>
```